scripts/analyze_codebase.py

```
@@ -0,0 +1,284 @@
+    
+    Returns list of (start_line, end_line, normalized_code) tuples.
+    """
+    blocks = []
+    try:
+        with open(filepath, "r", encoding="utf-8") as f:
+            content = f.read()
+        
+        tree = ast.parse(content)
+        
+        for node in ast.walk(tree):
+            if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef)):
```

Ruff (0.12.2)

Use `X | Y` in `isinstance` call instead of `(X, Y)`

Convert to `X | Y`

(UP038)